webui/react/src/stores/agents.tsx

```diff
@@ -0,0 +1,100 @@
+
+type AgentContext = {
+  agents: Loadable<Agent[]>;
+  updateAgents: (a: Loadable<Agent[]>) => void;
+};
+
+const initResourceTally: ClusterOverviewResource = { allocation: 0, available: 0, total: 0 };
+// TODO: dont export
+export const initClusterOverview: ClusterOverview = {
+  [ResourceType.CPU]: clone(initResourceTally),
+  [ResourceType.CUDA]: clone(initResourceTally),
+  [ResourceType.ROCM]: clone(initResourceTally),
+  [ResourceType.ALL]: clone(initResourceTally),
+  [ResourceType.UNSPECIFIED]: clone(initResourceTally),
+};
+
+const AgentsContext = createContext<AgentContext>({ agents: NotLoaded, updateAgents: noOp });
+
+export const AgentsProvider: React.FC<PropsWithChildren> = ({ children }) => {
+  const [state, setState] = useState<Loadable<Agent[]>>(NotLoaded);
+  return (
+    <AgentsContext.Provider value={{ agents: state, updateAgents: setState }}>
+      {children}
+    </AgentsContext.Provider>
+  );
+};
+
+export const useFetchAgents = (canceler: AbortController): (() => Promise<void>) => {
+  const { updateAgents } = useContext(AgentsContext);
+
+  return useCallback(async (): Promise<void> => {
+    try {
+      const agents = await getAgents({ signal: canceler.signal });
+      updateAgents(Loaded(agents));
+    } catch (e) {
+      handleError(e);
+    }
+  }, [canceler, updateAgents]);
+};
+
+export const useEnsureAgentsFetched = (canceler: AbortController): (() => Promise<void>) => {
+  const { agents, updateAgents } = useContext(AgentsContext);
+
+  return useCallback(async (): Promise<void> => {
+    if (agents !== NotLoaded) return;
+    try {
+      const agents = await getAgents({ signal: canceler.signal });
+      updateAgents(Loaded(agents));
+    } catch (e) {
+      handleError(e);
+    }
+  }, [canceler, updateAgents, agents]);
+};
+export const useAgents = (): Loadable<Agent[]> => {
+  // TODO: check undefined
+  const { agents } = useContext(AgentsContext);
+
+  return agents;
+};
+
+export const useClusterOverview = (): Loadable<ClusterOverview> => {
+  // Deep clone for render detection.
+  const agents = useAgents();
+  // TODO: memoize
+  return Loadable.map(agents, (agents) => {
+    const overview: ClusterOverview = clone(initClusterOverview);
+
+    agents.forEach((agent) => {
+      agent.resources
+        .filter((resource) => resource.enabled)
+        .forEach((resource) => {
+          const isResourceFree = resource.container == null;
+          const availableResource = isResourceFree ? 1 : 0;
+          overview[resource.type].available += availableResource;
+          overview[resource.type].total++;
+          overview[ResourceType.ALL].available += availableResource;
+          overview[ResourceType.ALL].total++;
+        });
+    });
+
+    for (const key in overview) {
+      const rt = key as ResourceType;
+      overview[rt].allocation =
+        overview[rt].total !== 0
+          ? percent((overview[rt].total - overview[rt].available) / overview[rt].total)
+          : 0;
+    }
```

Comment:
Yeah, the reason you don't want to have another store is because it's derived state. The agents list is the _source of truth_ for this data. Think of it the same way you think of a database, you want to avoid having the same information in multiple places because they could get out of sync.

Reply:
i was thinking to have in the same store, just as a different part of it. 

analogy to database would be that you have a materialized view (derived state) that is kept in sync with the underlying data and then accessed frequently.

Reply:
I would say that `useMemo` is the analogue for a materialized view.

Reply:
in this case for `useMemo`, though, the view is materialized separately for each client consuming it.

Reply:
That's specific to this implementation but it needn't be. We could put the `useMemo` in the context provider.

Reply:
Sorry, misunderstood you. I think that's a good idea if the computation is expensive. I think here it's probably not worth the extra complexity because the computation is so cheap (unless people have a lot more agents than I'm imagining)

Reply:
fair enough. agree on the cheapness of the computation.